### ISSUE TITLE
just saying...

### ISSUE THREAD

**Issue by RafalSkolasinski** [miphreal/python-rofi-menu#1]
...that this is awesome! ; - )

**Comment by RafalSkolasinski** [comment]
Could use an example how selection of menu entry could run some simple python function (defined directly in example.py) - if this is possible to achieve :)

**Comment by miphreal** [comment]
@RafalSkolasinski,
yep, it's possible to run any python code and control the behaviour on menu item selection

```python
#!/home/user/pyenv/versions/rofi/bin/python
import asyncio
from datetime import datetime
import os

import rofi_menu
from rofi_menu.menu import Operation


class OutputSomeTextItem(rofi_menu.Item):
    """Output arbitrary text on selection"""
    async def on_select(self, item_id, meta):
        # any python code
        await asyncio.sleep(0.1)
        return Operation(rofi_menu.OP_OUTPUT, (
            "💢 simple\n"
            "💥 multi-\n"
            "💫 <b>line</b>\n"
            "💣 <i>text</i>\n"
        ))


class DoAndExitItem(rofi_menu.Item):
    """Do something and exit"""
    async def on_select(self, item_id, meta):
        os.system('notify-send msg')
        return Operation(rofi_menu.OP_EXIT)


class CurrentDatetimeItem(rofi_menu.Item):
    """Show current datetime inside menu item"""
    async def load(self, meta):
        self.state = datetime.now().strftime('%A %d. %B %Y (%H:%M:%S)')

    async def render(self, meta):
        return f"🕑 {self.state}"


class CounterItem(rofi_menu.Item):
    """Increment counter on selection"""
    async def load(self, meta):
        await super().load(meta)
        self.state = self.state or 0

    async def on_select(self, item_id, meta):
        self.state += 1
        return await super().on_select(item_id, meta)

    async def render(self, meta):
        return f"🏃 Selected #{self.state} time(s)"


main_menu = rofi_menu.Menu(
    prompt="menu",
    items=[
        OutputSomeTextItem('Output anything'),
        DoAndExitItem('Do something and exit'),
        CurrentDatetimeItem(),
        CounterItem(),
    ],
)


if __name__ == "__main__":
    rofi_menu.run(main_menu)
```
If you run this example, you'll get
![image](https://user-images.githubusercontent.com/1089024/88265991-1a547900-ccd7-11ea-8a36-3e4662e76fc1.png)
On choosing `Output anything`
![image](https://user-images.githubusercontent.com/1089024/88266081-43750980-ccd7-11ea-88e1-18e8ed907531.png)


Thanks for bringing this up, I'll do some improvements to the lib to simplify certain cases
- [ ] Docs: describe how to execute arbitrary python code / handle selection
- [ ] Docs: describe rofi_menu.Item/Menu/NestedMenu API and add proper docstrings
- [x] Code: import `Operation` in the root `__init__.py` so it can be reached via `rofi_menu.Operation`
- [ ] Code: add `rofi_menu.contrib.python.PyFuncItem` to cover simple cases of running python code on selection (e.g. `PyFuncItem(on_select=do_some_stuff)`)

**Comment by RafalSkolasinski** [comment]
Thanks @miphreal!

Actually looking into the code of `ShellItem` class is also quite helpful.

One thing still escapes me: how could I make use of user's input properly. Let's say in one menu I want to have a list of some predefined options but also allow user to provide a new one. 

I figured out that I could have for example
```python
class Command(Item):
    def __init__(self, text=None, option=None, **kwargs):
        self.option = option
        super().__init__(text, **kwargs)

    async def on_select(self, item_id, meta):
        logging.info(f"Selected option: {self.option}")
        return Operation(rofi_menu.OP_EXIT)


def main():
    options = ["a", "b", "c", "d"]
    rofi_menu.run(
        rofi_menu.Menu(
            prompt="choose option", items=[Command(option=option, text=option) for option in options]
        )
    )
```

But how could I accept in additions to predefined `options` any user's input? 
Also probably creating multiple `Command` instances is most likely not the most optimal approach...

**Comment by RafalSkolasinski** [comment]
Just to give some context, I'd like to write a `workspace` switcher that first fetches already existing workspaces in `i3` and then allow to switch focus to one of them but if new option is provided it just creates a new workspace. The `i3ipc` part I have all worked out - just want to add a nice menu ; - ).

Happy to contribute example once it will be working properly.

**Comment by miphreal** [comment]
> But how could I accept in additions to predefined options any user's input?
> Also probably creating multiple Command instances is most likely not the most optimal approach...

In general, the script where we define rofi menu is called by rofi as `/path-to-example.py "user input or selected text"`. That means we can get user input from `sys.argv[1]` (tho it's a quite indirect way to do it).

I've just released `0.3.0` which ships raw user input in `meta.raw_input` (perhaps, it'll simplify some logic).


> Just to give some context, I'd like to write a workspace switcher that first fetches already existing workspaces in i3 and then allow to switch focus to one of them but if new option is provided it just creates a new workspace. The i3ipc part I have all worked out - just want to add a nice menu ; - ).
> 
> Happy to contribute example once it will be working properly.

I played locally to implement the logic you described (I added `generate_menu_items` in `0.3.0`, so it can be overridden to generate dynamic items). It might look as

```python
#!/home/user/.pyenv/versions/rofi/bin/python
import rofi_menu


class WorkspaceItem(rofi_menu.Item):
    def __init__(self, workspace, i3, **kwargs):
        self.workspace = workspace
        self.i3 = i3

        flags = kwargs.get('flags', set())
        if workspace.focused:
            flags.add(rofi_menu.FLAG_STYLE_ACTIVE)

        kwargs['text'] = workspace.name
        kwargs['flags'] = flags
        super().__init__(**kwargs)

    def clone(self):
        return WorkspaceItem(self.workspace, self.i3)

    async def on_select(self, item_id, meta):
        await self.i3.command(f"workspace {self.workspace.name}")
        return rofi_menu.Operation(rofi_menu.OP_EXIT)


class I3Workspaces(rofi_menu.Menu):
    prompt = 'workspaces'

    async def generate_menu_items(self, prefix_path, meta):
        from i3ipc.aio import Connection

        self.i3 = await Connection(auto_reconnect=True).connect()
        self.workspaces = await self.i3.get_workspaces()

        return [(w.name, WorkspaceItem(workspace=w, i3=self.i3)) for w in self.workspaces]

    async def on_select(self, item_id, meta):
        if not item_id and meta.raw_input:
            await self.i3.command(f'workspace {meta.raw_input}')
            return rofi_menu.Operation(rofi_menu.OP_EXIT)

        return await super().on_select(item_id, meta)


if __name__ == "__main__":
    rofi_menu.run(I3Workspaces())
```

![image](https://user-images.githubusercontent.com/1089024/88411978-86230880-cde1-11ea-8c6c-08bc04cbfc22.png)

**Comment by RafalSkolasinski** [comment]
This is so cool, really good stuff. I still have a few questions, mostly curiosity. 

1. Why did you decide to go with `async` implementation? I do not see opportunity for many concurrent operations happening. Do I miss some potential use case where it can make a difference? 

2. As rofi passes arguments to the script when calling as `sys.argv[1]` does it mean that script is independently executed for each submenu user enters?

**Comment by miphreal** [comment]
@RafalSkolasinski,

> Why did you decide to go with async implementation? I do not see opportunity for many concurrent operations happening. Do I miss some potential use case where it can make a difference?

Just wanted to cover all probable cases related to doing http requests or any io operations to load initial states for menu items (e.g network stats "ping X ms", health checks for servers/nodes, or any other automation requiring multiple network requests). doing those sequentially would take much longer.
Anyway, nothing prevents you to write non-async blocking code within `async def ...` method.

> As rofi passes arguments to the script when calling as sys.argv[1] does it mean that script is independently executed for each submenu user enters?

Basically, rofi executes the script each time we select menu item. To handle submenus and understand which menu item was selected, the lib attaches some meta to each menu item when outputs menu for rofi and then, when rofi passes the selected menu item back to the script - it extracts that meta, finds a proper menu item in the tree of menu items and then a concrete menu item handles selection. _[not sure if I described it clearly]_

**Comment by RafalSkolasinski** [comment]
Thank you @miphreal, that's clear enough :+1: 

Regarding the `raw_input` feature a good example into docs could me something involving user name.
Could be `Change User (current user XYZ)` that opens menu to choose one from few prepared options or provide a new one.

About the workspace changer - your example works pretty well!
